lib/ocpp/v201/smart_charging.cpp

```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
+    int current_period_limit = MAX_PERIOD_LIMIT;
+    LimitStackLevelPair significant_limit_stack_level_pair = {MAX_PERIOD_LIMIT, -1};
+
+    // calculate every ChargingSchedulePeriod of result within this while loop
+    while (SmartChargingHandler::within_time_window(end_time, temp_time)) {
+        // while (duration_cast<seconds>(end_time.to_time_point() - temp_time.to_time_point()).count() > 0) {
+
+        std::map<ChargingProfilePurposeEnum, LimitStackLevelPair> current_purpose_and_stack_limits =
+            get_initial_purpose_and_stack_limits(); // this data structure holds the current lowest limit and stack
+                                                    // level for every purpose
+        ocpp::DateTime temp_period_end_time;
+        int32_t temp_number_phases;
+
+        for (const ChargingProfile& profile : valid_profiles) {
+            EVLOG_info << "ProfileId #" << profile.id << " Kind: " << profile.chargingProfileKind;
+
+            if (profile.stackLevel > current_purpose_and_stack_limits.at(profile.chargingProfilePurpose).stack_level) {
+                // EVLOG_info << "calculate_composite_schedule> profile.stackLevel > current_purpose_and_stack_limits";
+                EVLOG_info << "boop";
+
+                // this data structure holds the respective period and period end time for temp_time point in time
+                const PeriodDateTimePair period_date_time_pair = this->find_period_at(temp_time, profile, evse_id);
+
+                if (period_date_time_pair.period.has_value()) {
+                    const ChargingSchedulePeriod period = period_date_time_pair.period.value();
+
+                    temp_period_end_time = period_date_time_pair.end_time;
+
+                    if (period.numberPhases) {
+                        temp_number_phases = period.numberPhases.value();
+                    } else {
+                        temp_number_phases = DEFAULT_AND_MAX_NUMBER_PHASES;
+                    }
+
+                    ChargingSchedule first_charging_schedule = profile.chargingSchedule.front();
+
+                    float limit =
+                        get_power_limit(period.limit, temp_number_phases, first_charging_schedule.chargingRateUnit);
+                    int32_t stackLevel = profile.stackLevel;
+
+                    EVLOG_info << "period.has_value() limit = " << limit;
+                    EVLOG_info << "period.has_value() stackLevel = " << limit;
+
+                    // update data structure with limit and stack level for this profile
+                    current_purpose_and_stack_limits.at(profile.chargingProfilePurpose).limit = limit;
+                    current_purpose_and_stack_limits.at(profile.chargingProfilePurpose).stack_level = stackLevel;
+                } else {
+                    // skip profiles with a lower stack level if a higher stack level already has a limit for this
+                    // point in time
+                    continue;
+                }
```

Comment:
This `continue` doesn't seem necessary here - we're at the end of the loop already, so we're already going to continue in either case.